components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+                             -> Result<Vec<PackageArchive>>
+        where T: UIWriter
+    {
+        let mut downloaded_artifacts = Vec::<PackageArchive>::new();
+
+        ui.status(Status::Downloading,
+                  format!("Downloading {} artifacts", expanded_idents.len()))?;
+
+        for (ident, target) in expanded_idents {
+            let archive: PackageArchive = match self.get_downloaded_archive(ui, ident, *target) {
+                Ok(v) => v,
+                Err(e) => {
+                    // Is this the right status? Or should this be a debug message?
+                    ui.status(Status::Missing,
+                              format!("Error fetching archive {} for {}: {:?}",
+                                      ident, *target, e))?;
+                    return Err(e);
+                }
+            };
+
+            downloaded_artifacts.push(archive);
+        }
+
+        Ok(downloaded_artifacts)
+    }
+
+    fn determine_latest_from_ident<T>(&self,
+                                      ui: &mut T,
+                                      ident: &PackageIdent,
+                                      target: PackageTarget)
+                                      -> Result<Package>
+        where T: UIWriter
+    {
+        // Unlike in the install command, we always hit the online
+        // depot; our purpose is to sync with latest, and falling back
+        // to a local package would defeat that. Find the latest
+        // package in the proper channel from Builder API,
+        ui.status(Status::Determining,
+                  format!("latest version of {} for {} in the '{}' channel",
+                          ident, target, self.channel))?;
+        match self.fetch_latest_package_in_channel_for(ident, target, self.channel, self.token) {
+            Ok(latest_package) => {
+                ui.status(Status::Using,
+                          format!("{} as latest matching {} for {}",
+                                  latest_package.ident, ident, target))?;
+                Ok(latest_package)
+            }
+            Err(Error::APIClient(APIError(StatusCode::NOT_FOUND, _))) => {
+                // In install we attempt to recommend a channel to look in. That's a bit of a
+                // heavyweight process, and probably a bad idea in the context of
+                // what's a normally a batch process. It might be OK to fall back to
+                // the stable channel, but for now, error.
+                ui.warn(format!("No packages matching ident {} for {} exist in the '{}' \
+                                 channel. Check the package ident, target, channel and builder \
+                                 url for correctness",
+                                ident, target, self.channel))?;
+                Err(Error::from(CommonError::PackageNotFound(
+                    format!("{} for {} in channel {}", ident, target, self.channel).to_string(),
```

Comment:
The `format` macro returns a `String` already
```suggestion
                    format!("{} for {} in channel {}", ident, target, self.channel),
```